web_src/js/features/heatmap.js

```diff
@@ -7,8 +7,15 @@ export default async function initHeatmap() {
   if (!el) return;
 
   try {
-    const values = JSON.parse(el.dataset.heatmapData).map(({contributions, timestamp}) => {
-      return {date: new Date(timestamp * 1000), count: contributions};
+    const heatmap = {};
+    JSON.parse(el.dataset.heatmapData).forEach(({contributions, timestamp}) => {
+      // Convert to user timezone and sum contributions by date
+      const dateStr = new Date(timestamp * 1000).toDateString();
+      heatmap[dateStr] = (heatmap[dateStr] || 0) + contributions;
+    });
+
+    const values = Object.keys(heatmap).map((v) => {
+      return {date: new Date(v), count: heatmap[v]};
     });
```

Comment:
Why iterate twice?

Reply:
The user-heatmap data is made up of timestamps (at 15-minute intervals) to contributions, ActivityHeatmap.vue accepts daily contributions.

The first loop converts the timestamps to the users timezone aware date string (stripping time) and sums the contributions.
The second loop converts it back to a date object with the proper format (list of objects).

Reply:
Pretty sure it can be all done in one iteration (e.g. in the previous `map` callback). The only reason to loop twice through the data would be to deduplicate it, but I think that is not necessary.

Reply:
Try this:

```js
const values = JSON.parse(el.dataset.heatmapData).map(({contributions, timestamp}) => {
  // Convert to user timezone and sum contributions by date
  return {
    date: new Date(timestamp * 1000).toDateString(),
    count: (heatmap[dateStr] || 0) + contributions
  }
});

Reply:
Note, time will probably be in incorrect timezone, we'd need a helper function to convert to current timezone, like this (untested):

```js
function convertToLocaleTimezone = (date) => {
  const offset = (new Date()).getTimezoneOffset() * -60000;
  return new Date(date.getTime() + offset);
};
```

If that works, put it in `web_src/js/utils.js` and import it in this file.

Reply:
I tried it and it does not work because ActivityMap.vue will not do any summation, it will only pick one entry per day. You have to de-duplicate it as you've mentioned.

Reply:
Yes, graph expects absolute per-day values. I will test my suggestion above later.

Reply:
The timezone has been correct in my testing. new Date() in JavaScript will return a date with the local browsers timezone automatically. I have also tested it with changing my timezone using a browser extension.

Reply:
@silverwind  no, do not try to correct for the timezone.
Unix time, which is stored in the database is based on UTC.
When you call `new Date( int milliseconds )`  the parameter is defined as the ECMA epoch which is also UTC based.
- <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#parameters>
- <https://tc39.es/ecma262/#sec-date-objects>

Here's the trick, when the code calls `toDateString()` it is the responsibility of the javascript engine to provide the time using the local system timezone.

- <https://tc39.es/ecma262/#sec-date.prototype.todatestring>

This means it will show the correct information for the user in their timezone.  If I compare what it looks like for me with someone on the other side of the world using the same server, yes the heatmap will look different.  I think that's OK since it is easy to explain.

This is a user heatmap of user contribution and a user is likely to be looking the page from their computer mapped to their local timezone .  It is also clear when you hover over a box what frame of reference is being used, and I think this is far more intuitive.

Reply:
Thanks for clearing that up. I wasn't aware of this built-in timezone conversion of the various `Date()` `.toString` methods.

I think we also need to clean up the click mechanism which (wrongly) converts from UTC to local offset:

https://github.com/go-gitea/gitea/blob/c680eb2cc74823f017f5ad5a4c9e8094cde311f4/web_src/js/components/ActivityHeatmap.vue#L58-L59

Reply:
I can't explain how, but in my testing clicking the days on the heatmap to filter the actions works as intended for me in EST.
Here are some screenshots:

![image](https://user-images.githubusercontent.com/584014/114336569-4b9efe80-9b1d-11eb-9f54-b13d8ccf28bd.png)
![image](https://user-images.githubusercontent.com/584014/114336649-6cffea80-9b1d-11eb-94ea-afd51348f92d.png)

Reply:
I think that's because when you create a new date from a string rather than a millisecond epoch, again the date object is responsible for converting that to utc which is what the date object stores.  I'd have to double check to be 100% sure but that's what I remember.